### DISCUSSION TITLE
How do I make the markdown preview render images sourced from custom schemes?

### DISCUSSION THREAD

**Discussion by Tommycore** [microsoft/vscode-discussions#113] [Extension Development QnA] [answered]
Short overview:
- I have an extension that implements a custom file scheme.
- When running my extension I have all other extensions turned off by default.
- My automated tests prove that files loaded via that scheme do load properly and yield the same results as when said files are loaded via the standard file scheme.
- In my markdown, I want to be able to read files over my custom scheme. That does not work.
- Specifying a scheme outside http(s) yields no result, not even when naming the standard file scheme.

The picture below shows a side by side view of 5 attempts to embedd pictures in markdown:
1. My custom scheme
2. The file scheme but with the file scheme specified
3. A non-existant scheme
4. Https scheme
5. No scheme specified

As you can see, 4 and 5 work. 2 doesn't work even if I enter the absolute path to the picture, encoded or not. At least when I enter the absolute path encoded I can Ctrl-Click in the texteditor and VS Code opens up the the picture in a new tab. As soon as the uri begins with `file:` the markdown renderer turns into text and no longer accepts the line as an embedded image. Here's a [link](https://github.com/Tommycore/Smithy/blob/84122e0ef1832f9dba036bf1caff19d0b6bf23e3/src/AssetSourcer.ts#L53) to the method from my FileSystemProvider that is supposed to load the image. What am I doing wrong that 1 doesn't work? Will I have to implement a custom MarkdownIt plugin that interprets uris with my custom scheme?

![image](https://user-images.githubusercontent.com/19580006/186880991-6854d7ec-5024-40ef-825e-a5958173cf56.png)

**Comment by isidorn** [comment]
Maybe @mjbvz knows the answer

**Comment by Tommycore** [comment]
Could it maybe have something to do with absolute and relative paths? I was running into a problem with another file scheme, where paths didn't get resolved when they didn't start with `/` and used an authority. Apparently VS Code was trying to interpret `my/random/path` as a relative path and got confused as to relative to *what* the path was. To the point that I didn't even hit any breakpoints in my file scheme provider. Once I started using an authority in the uri and had my path begin with a `/` everything went fine. Wasn't able to use this to resolve the problem from this thread though =/

**Comment by mjbvz** [comment, marked as answer]
Can you please share a minimal extension that demonstrates this problem? That will help me understand if there is a bug in your code or in something VS Code is doing

**Reply by Tommycore** [reply]
It's been a while and I haven't worked with that specific problem for over 4 month now. I tried reconstructing what I could from memory: https://github.com/Tommycore/markdownpreviewimagefromcustomschemedemo

Interesting enough, I don't seem to even approach using the file system provider in this case. Because if I would, there'd be a [popup](https://github.com/Tommycore/markdownpreviewimagefromcustomschemedemo/blob/main/src/AssetFileSystemProvider.ts#L37) in the bottom right. Like I said - it's been a while. Hope it still helps.

**Reply by pcj** [reply]
If you work on this again, consider https://github.com/microsoft/vscode/issues/171393#issuecomment-1386080735.  I have observed in different contexts that a custom filesystem must be registered as a folder in the workspace to resolve properly. I see in your extension that you never call `vscode.workspace.updateWorkspaceFolders` with your custom scheme.

**Reply by Tommycore** [reply]
Thanks, I'll keep that in mind. I must admit though that it is unlikely that I'll try again anytime soon.
